src/model/guild/automod.rs

```diff
@@ -0,0 +1,859 @@
+}
+
+#[derive(Deserialize, Serialize)]
+#[serde(rename = "ActionMetadata")]
+struct Timeout {
+    #[serde(rename = "duration_seconds")]
+    duration: u64,
+}
+
+impl<'de> Deserialize<'de> for Action {
+    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        #[derive(Deserialize)]
+        #[serde(field_identifier, rename_all = "snake_case")]
+        enum Field {
+            Type,
+            Metadata,
+            Unknown(String),
+        }
+
+        struct Visitor;
+
+        impl<'de> serde::de::Visitor<'de> for Visitor {
+            type Value = Action;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+                formatter.write_str("automod rule action")
+            }
+
+            fn visit_map<A: MapAccess<'de>>(self, mut map: A) -> Result<Self::Value, A::Error> {
+                let mut kind = None;
+                let mut metadata = None;
+
+                while let Some(key) = map.next_key()? {
+                    match key {
+                        Field::Type => {
+                            if kind.is_some() {
+                                return Err(Error::duplicate_field("type"));
+                            }
+                            kind = Some(map.next_value()?);
+                        },
+                        Field::Metadata => {
+                            if metadata.is_some() {
+                                return Err(Error::duplicate_field("metadata"));
+                            }
+                            metadata = Some(map.next_value::<Value>()?);
+                        },
+                        Field::Unknown(_) => {
+                            map.next_value::<IgnoredAny>()?;
+                        },
+                    }
+                }
+                let kind = kind.ok_or_else(|| Error::missing_field("type"))?;
+                match kind {
+                    ActionType::BlockMessage => Ok(Action::BlockMessage),
+                    ActionType::Alert => {
+                        let alert: Alert = metadata
+                            .ok_or_else(|| Error::missing_field("metadata"))?
+                            .deserialize_into()
+                            .map_err(DeserializerError::into_error)?;
+                        Ok(Action::Alert(alert.channel_id))
+                    },
+                    ActionType::Timeout => {
+                        let timeout: Timeout = metadata
+                            .ok_or_else(|| Error::missing_field("metadata"))?
+                            .deserialize_into()
+                            .map_err(DeserializerError::into_error)?;
+                        Ok(Action::Timeout(timeout.duration))
+                    },
+                    ActionType::Unknown(unknown) => Ok(Action::Unknown(unknown)),
+                }
+            }
+        }
+
+        deserializer.deserialize_any(Visitor)
+    }
+}
+
+impl Serialize for Action {
```

Comment:
You know the drill

If there's a good reason, it would be nice to have it commented above the impls

Reply:
Nah its fine